R/buffer.R

```diff
@@ -0,0 +1,87 @@
+#' Apply an inclusion radius and exclusion buffer to indices
+#'
+#' @param data An object of class `sf` or `sfc`.
+#' @param indices List of indices in each fold generated by `split_unnamed()`.
+#' @param radius Numeric: points within this distance of the initially-selected
+#' test points will be assigned to the assessment set. If `NULL`, no radius is
+#' applied.
+#' @param buffer Numeric: points within this distance of any point in the
+#' test set (after `radius` is applied) will be assigned to neither the analysis
+#' or assessment set. If `NULL`, no buffer is applied.
+#'
+#' @keywords internal
+buffer_indices <- function(data, indices, radius, buffer, call = rlang::caller_env()) {
+
+  if (identical(sf::st_crs(data), sf::NA_crs_)) {
+    rlang::abort(
+      c("`buffer` and `radius` require `data` to have a non-NA coordinate reference system",
+        "i" = "Set the CRS for your data using `sf::st_set_crs()`"
+      ),
+      call = call
+    )
+  }
+
+  if (sf::st_is_longlat(data) && !sf::sf_use_s2()) {
+    rlang::abort(
+      c(
+        "`buffer` and `radius` can only be used with geographic coordinates when using the s2 geometry library",
+        "i" = "Reproject your data into a projected coordinate reference system using `sf::st_transform()`",
+        "i" = "Or install the `s2` package and enable it using `sf::sf_use_s2(TRUE)`"
+      ),
+      call = call
+    )
+  }
+
+  n <- nrow(data)
+
+  if (is.null(radius)) radius <- 0
+  if (is.null(buffer)) buffer <- 0
+
+  if (radius) {
```

Comment:
I'm worried about using `if()` like this on a floating point number. Maybe `identical()` is better? Like `if(!identical(radius, 0L)`? Or can we rewrite this to avoid this kind of comparison?

Reply:
So there's two challenges here:

1. I think this is incompatible with https://github.com/tidymodels/spatialsample/pull/39/#discussion_r890495535 (Hannah's comment). If we don't check for a float 0 (which is what the user is probably providing), then running `row_ids_within_dist()` as it's currently written will set `indices` to a list of `integer(0)`, which is wrong. 
2. At the same time, I believe that running `row_ids_within_dist` for a distance of 0 _without_ that check will return all polygons that touch `indices` -- a polygon sharing a border with another is within distance 0 of its neighbor. 

Something that might _feel_ better is changing this to:

```r
if (is.null(radius)) radius <- FALSE
if (is.null(buffer)) buffer <- 0

if (radius) {
```

But if the user is setting `radius` to 0, then this is still running a conditional on a float. 

If we wanted, we could change the block to:

```r
if (is.null(buffer)) buffer <- 0

if (identical(radius, 0) || identical(radius, 0L) || is.null(radius)) {
```

Does that feel better?

Reply:
I think by writing this out, I convinced myself the last chunk _does_ feel better (:rubber: :duck:) and am going to push that up as a commit